aspnetcore/signalr/hub-filters.md

```diff
@@ -0,0 +1,193 @@
+
+If a filter wants to skip a hub method invocation we recommend throwing a `HubException` and not calling `next` so the client will receive an error if it was expecting a result.
+
+## Using hub filters
+
+When writing the filter logic, try to make it generic by using attributes on hub methods instead of checking for hub method names.
+
+For example, lets write a filter that will check an argument to a hub method for banned phrases and replace any phrases it finds with "***".
+For the purposes of this example we'll assume a `LanguageFilterAttribute` class is defined and it has a property named `FilterArgument` that can be set when using the attribute.
+
+First, place the attribute on the hub method that has a string argument that we want to clean up.
+```csharp
+public class ChatHub
+{
+    [LanguageFilter(filterArgument: 0)]
+    public async Task SendMessage(string message, string username)
+    {
+        await Clients.All.SendAsync("SendMessage", $"{username} says: {message}");
+    }
+}
+```
+
+Next, define a hub filter that will check for the attribute and replace banned phrases in an argument to the hub method with "***".
+```csharp
+public class LanguageFilter : IHubFilter
+{
+    // populated from a file or inline
+    private List<string> bannedPhrases = new List<string>();
+
+    public async ValueTask<object> InvokeMethodAsync(HubInvocationContext invocationContext, Func<HubInvocationContext, ValueTask<object>> next)
+    {
+        var languageFilter = (LanguageFilterAttribute)Attribute.GetCustomAttribute(invocationContext.HubMethod, typeof(LanguageFilterAttribute));
+        if (invocationContext.HubMethodArguments.Count > languageFilter.FilterArgument &&
+            invocationContext.HubMethodArguments[languageFilter.FilterArgument] is string str)
+        {
+            foreach (var bannedPhrase in bannedPhrases)
+            {
+                str = str.Replace(bannedPhrase, "***");
+            }
+            if (invocationContext.HubMethodArguments is object[] arguments)
+            {
+                arguments[languageFilter.FilterArgument] = str;
+            }
+            else
+            {
+                arguments = invocationContext.HubMethodArguments.ToArray();
+                arguments[languageFilter.FilterArgument] = str;
+                invocationContext = new HubInvocationContext(invocationContext.Context, invocationContext.ServiceProvider, invocationContext.Hub, invocationContext.HubMethod, arguments);
+            }
+        }
+
+        return await next(invocationContext);
+    }
+}
+```
+
+Finally, we need to register the hub filter. Because this filter relies on state outside of the normal hub method scope we need to register it as a singleton.
```

Comment:
What state?

Reply:
The list of banned phrases

Reply:
We should make that more explicit.

```suggestion
Finally, we need to register the hub filter. To avoid reinitializing the banned phrases list for every invocation, we register the hub filter as a singleton.
```